postprocess/postprocess.go

```diff
@@ -0,0 +1,231 @@
+*/
+func UpdatePointCloud(
+	data []byte,
+	updatedData *[]byte,
+	tasks []Task,
+) error {
+	*updatedData = append(*updatedData, data...)
+
+	// iterate through tasks and add or remove points
+	for _, task := range tasks {
+		switch task.Instruction {
+		case Add:
+			err := updatePointCloudWithAddedPoints(updatedData, task.Points)
+			if err != nil {
+				return err
+			}
+		case Remove:
+			err := updatePointCloudWithRemovedPoints(updatedData, task.Points)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func updatePointCloudWithAddedPoints(updatedData *[]byte, points []r3.Vector) error {
+	reader := bytes.NewReader(*updatedData)
+	pc, err := pointcloud.ReadPCD(reader)
+	if err != nil {
+		return err
+	}
+
+	for _, point := range points {
+		/*
+			Viam expects pointcloud data with fields "x y z" or "x y z rgb", and for
+			this to be specified in the pointcloud header in the FIELDS entry. If color
+			data is included in the pointcloud, Viam's services assume that the color
+			value encodes a confidence score for that data point. Viam expects the
+			confidence score to be encoded in the blue parameter of the RGB value, on a
+			scale from 1-100.
+		*/
+		err := pc.Set(point, pointcloud.NewColoredData(color.NRGBA{B: fullConfidence, R: math.MaxUint8}))
+		if err != nil {
+			return err
+		}
+	}
+
+	var buf bytes.Buffer
+	err = pointcloud.ToPCD(pc, &buf, pointcloud.PCDBinary)
+	if err != nil {
+		return err
+	}
+
+	// Initialize updatedData with new points
+	*updatedData = make([]byte, buf.Len())
+	updatedReader := bytes.NewReader(buf.Bytes())
+	_, err = updatedReader.Read(*updatedData)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func updatePointCloudWithRemovedPoints(updatedData *[]byte, points []r3.Vector) error {
+	reader := bytes.NewReader(*updatedData)
+	pc, err := pointcloud.ReadPCD(reader)
+	if err != nil {
+		return err
+	}
+
+	updatedPC := pointcloud.NewWithPrealloc(pc.Size() - len(points))
+	pointsVisited := 0
+
+	filterRemovedPoints := func(p r3.Vector, d pointcloud.Data) bool {
+		pointsVisited++
+		// Always return true so iteration continues
+
+		for _, point := range points {
+			// remove all points within the removalRadius from the removed points
+			if point.Distance(p) <= removalRadius {
+				return true
+			}
+		}
+
+		err := updatedPC.Set(p, d)
+		// end early if point cannot be set
+		return err == nil
+	}
+
+	pc.Iterate(0, 0, filterRemovedPoints)
+
+	// confirm iterate did not have to end early
+	if pc.Size() != pointsVisited {
+		/*
+			Note: this condition will occur if some error occurred while copying valid points
+			and will be how we can tell that this error occurred: err := updatedPC.Set(p, d)
+		*/
+		return ErrRemovingPoints
+	}
+
+	buf := bytes.Buffer{}
+	err = pointcloud.ToPCD(updatedPC, &buf, pointcloud.PCDBinary)
+	if err != nil {
+		return err
+	}
+
+	// Overwrite updatedData with new points
+	*updatedData = make([]byte, buf.Len())
```

Comment:
should we check if nil? What happens if it is?

Reply:
if what is nil?

Reply:
`*updatedData`, which we pass into Read. Come to think fo it can `buf.Len()` ever be zero?

Reply:
I tested this in fmt.Print(make([]byte, 0) == nil) and it it is false

Reply:
I think there's another one of these higher up, don't forget that one too!